source/common/http/header_map_impl.h

```diff
@@ -1,43 +1,127 @@
 #pragma once
 
+#include "headers.h"
+
 #include "envoy/http/header_map.h"
 
+#include "common/common/non_copyable.h"
+
 namespace Http {
 
 /**
- * Implementation of Http::HeaderMap. This implementation uses a linked list to keep the headers
- * in order. If it's proven that having fast lookup on top is helpful we can add a map referencing
- * capability as well.
+ * These are definitions of all of the inline header access functions described inside header_map.h
+ */
+#define DEFINE_INLINE_HEADER_FUNCS(name)                                                           \
+public:                                                                                            \
+  const HeaderEntry* name() const override { return inline_headers_.name##_; }                     \
+  HeaderEntry* name() override { return inline_headers_.name##_; }                                 \
+  HeaderEntry& insert##name() override {                                                           \
+    return maybeCreateInline(&inline_headers_.name##_, Headers::get().name);                       \
+  }                                                                                                \
+  void remove##name() override { removeInline(&inline_headers_.name##_); }
+
+#define DEFINE_INLINE_HEADER_STRUCT(name) HeaderEntryImpl* name##_;
+
+/**
+ * Implementation of Http::HeaderMap. This is heavily optimized for performance. Roughly, when
+ * headers are added to the map, we do a hash lookup to see if it's one of the O(1) headers.
+ * If it is, we store a reference to it that can be accessed later directly. Most high performance
+ * paths use O(1) direct access. In general, we try to copy as little as possible and allocate as
+ * little as possible in any of the paths.
  */
 class HeaderMapImpl : public HeaderMap {
 public:
-  typedef std::pair<LowerCaseString, std::string> HeaderEntry;
-
-  HeaderMapImpl() {}
-  HeaderMapImpl(const std::initializer_list<HeaderEntry>& values) : headers_(values) {}
+  HeaderMapImpl();
+  HeaderMapImpl(const std::initializer_list<std::pair<LowerCaseString, std::string>>& values);
   HeaderMapImpl(const HeaderMap& rhs);
 
   /**
+   * Add a header via full move. This is the expected high performance paths for codecs populating
+   * a map when receiving.
+   */
+  void addViaMove(HeaderString&& key, HeaderString&& value);
+
+  /**
    * For testing. Equality is based on equality of the backing list.
```

Comment:
can you also put comments that order of headers in the list matters as it's position dependent comparison.